Add protection for empty waveforms

gondiaz opened the pull request:
This PR add a protection for empty waveforms in signal finder. 
I realized about this issue throught detsim simulation of background events close to the chamber edges at radious higher than the active radious. 
In such cases the waveforms created by detsim are empty, ie they are a list of zeros. Current implementation is not prepared to deal this case. Also, note that a `trigger_thr` parameter high enought, no signals are found and the same problem arises. This issue also affects buffy city.

gondiaz committed:
Add a test that demonstrates an unwanted behaviour

When an empty waveform is created in detsim the signal finder does
not find any pulse, breaking the city flow.

gondiaz committed:
Return empty list if no signal is found

gondiaz committed:
Add filter for events with no signal

gondiaz committed:
Rewrite detsim pipe with signal filter implementation

Avoid unnecessary branching in detsim

gondiaz committed:
Add empty waveforms test assertions

gondiaz committed:
Rewrite buffy pipe with signal filter implementation

jacg commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -59,6 +59,20 @@ def test_detsim_filter_active_hits(ICDATADIR, output_tmpdir):
 
 
+def test_detsim_filter_empty_waveforms(ICDATADIR, output_tmpdir):
+    # the first event radious is slighty above NEW active radious of 208.0 mm
```
Typo: radious -> radius (twice)

jacg commented on invisible_cities/detsim/buffer_functions.py:
```diff
@@ -55,6 +55,7 @@ def find_signal_start(wfs          : Union[pd.Series, np.ndarray],
+    if len(indices) == 0: return []
     ## Just using this and the stand_off for now
     ## taking first above sum threshold.
     ## !! To-do: make more robust with min int? or similar
```
It's really not worth opening a separate PR for it, so, while we're touching this file, could you please add a single commit which fixes all the `to-do`s in this file (and any others that you are already touching in this PR), changing them to this *precise* spelling: `TODO` (exactly four characters, all uppercase, no spaces, no dashes). This is because many development tools recognize `TODO` and can help managing them.

jacg commented on invisible_cities/cities/components.py:
```diff
@@ -878,6 +878,12 @@ def calculate_and_save_buffers(buffer_length    : float        ,
 
+    filter_events_signal = fl.map(lambda x: len(x) > 0,
+                                          args= 'pulses',
+                                          out = 'passed_signal')
```
Fix indentation

jacg commented on invisible_cities/cities/components.py:
```diff
@@ -902,10 +908,13 @@ def calculate_and_save_buffers(buffer_length    : float        ,
                                      saved_buffers                ))
 
     fn_list = (find_signal      ,
```
`fn_list` is a very unhelpful name: please use a name that summarizes what the sub-pipeline does.

jacg commented on invisible_cities/cities/components.py:
```diff
@@ -902,10 +908,13 @@ def calculate_and_save_buffers(buffer_length    : float        ,
 
     fn_list = (find_signal      ,
+               filter_events_signal,
+               fl.branch(write_signal_filter),
```
A `branch` containing a function wrapped in `sink` is a common pattern, which is why `spy` exists as an abstraction:

```python
spy(fn)
```

is a shorthand for 

```python
fn_wrapped_in_sink = sink(fn)
branch(fn_wrapped_in_sink)
```
or, more simply

```python
branch(sink(fn))
```

Apart from being slightly shorter to type, it sends a strong signal to the human reader about the simplicity of what is going on.

gondiaz replied:
The `sink`ed function, namely `write_signal_filter`, requires `args` which are not admitted in `spy`. I don't see how we can deal with it.

jacg replied:
Ah, OK, `sink` is one of the privileged few that have been endowed with argument picking. (Which is why [liquidata](https://github.com/jacg/liquidata) implements argument picking and result setting as independent, orthogonal and universally applicable utilities).

So, yes, in this case the sink-in-branch approach is probably the cleanest.

jacg commented on invisible_cities/cities/components.py:
```diff
@@ -902,10 +908,13 @@ def calculate_and_save_buffers(buffer_length    : float        ,
                calculate_buffers,
                order_sensors    ,
-               buffer_writer_   )
+               fl.branch(buffer_writer_))
```
Again, you could use `spy` here, and remove the `sink` wrapper in the definition of `buffer_writer_`

jacg commented on invisible_cities/cities/detsim.py:
```diff
@@ -213,9 +213,10 @@ def detsim(*, files_in, file_out, event_range, print_mod, compression,
                                          fl.branch("event_number"     ,
                                                    evtnum_collect.sink),
-                                         buffer_calculation),
+                                         fl.sink(lambda x: x)),
```
What is the point of this? Is it to have a `/dev/null` sink? (One that just throws the data away?) 

If so, then simply don't branch off at the previous step!

gondiaz replied:
We could sink the pipe inside `calculate_and_save_buffers`, but I wanted to keep the `evtnum_collect` counter. I dont know how to `sink` the main pipe afterwards so I added this dummy `sink` which is irrelevant for the city itself.

jacg replied:
I don't follow. At the moment you have

```
buffer_calculation ----> /dev/null
                  \
                    ---> pick event_number -> evtnum_collect.sink
```
Why can't you do the following instead?
```
buffer_calculation ---> pick event_number -> evtnum_collect.sink
```

gondiaz replied:
Yes, you are right. My bad because didn't really understand what `"event_number"` did inside a pipe.

jacg commented on invisible_cities/cities/buffy.py:
```diff
@@ -112,9 +112,10 @@ def buffy(files_in     , file_out   , compression      , event_range,
                                           fl.branch("event_number"      ,
                                                     evtnum_collect.sink),
-                                          buffer_calculation            )   ,
+                                          fl.sink(lambda x: x))         ,
```
Again, you won't need a cryptic `/dev/null` sink if you do the previous step without branching off.

jacg commented on invisible_cities/cities/components.py:
```diff
@@ -902,10 +908,13 @@ def calculate_and_save_buffers(buffer_length    : float        ,
 
     fn_list = (find_signal      ,
+               filter_events_signal,
+               fl.branch(write_signal_filter),
```
Generally speaking, lining the commas up vertically helps to avoid mistakes and misunderstandings, and this can be especially important in dataflow pipelines. Now, adjusting the commas at the end of the line is a PITA, and adds unnecessary noise to the diffs.

I recommend that we adopt the following style. Rather than

``` python
fn(foo,
   baaaaaaaa,
   baz,
   quux,
   quuuuuux)
```
where comma-errors or surprises are difficult to spot, or 
``` python
fn(foo      ,
   baaaaaaaa,
   baz      ,
   quux     ,
   quuuuuux )
```
which is a PITA to keep aligned, do this:
``` python
fn( foo
  , baaaaaaaa
  , baz
  , quux
  , quuuuuux)
```
This will really freak out the PEP8-ists, (and many of you) but, turst me, it's the best way forward :-)

gondiaz committed:
Cosmetics

Fix typo in test comment

Rename subpipe

Re-align commas in calculate_and_save_buffers

Re-align commas in detsim

Re-align commas in buffy

Rename TODOs in file

jacg commented on invisible_cities/cities/components.py:
```diff
@@ -898,24 +898,24 @@ def calculate_and_save_buffers(buffer_length    : float        ,
-                                           length_trk = nsamp_sipm,
-                                           max_subevt = max_subevt),
+    buffer_writer_    = sink(buffer_writer(  h5out
+                                           , run_number = run_number
```
I'm partial to aligning the commas directly under the opening bracket (and I'm pretty sure that is de-facto standard for this style), but I won't kick up a fuss if you leave it as it is.

mmkekic reviewed (approved):
This PR adds a filter in case no signal pulses were found in the waveforms. New filter is used both in detsim and buffy (generalizing  the previous buffy filter that was triggered for only all-0s waveforms). The test to show unwanted behavior in detsim is added - it fails before the changes and passes afterwards.
The code changed is simple and clean. Thanks!
